### PR TITLE
[EC-43] Block validation

### PR DESCRIPTION
**Pull request by AlanVerbner**
## Description

Block Holistic Validity based on Section 4.4.2 of http://paper.gavwood.com/ checked while doing fast Sync

### Considerations
- Task EC-73 was added in order to handle invalid blocks received while FastSync
- All validations except block state after execution is performed

### CODE REVIEW

File: src/main/scala/io/iohk/ethereum/network/FastSyncActor.scala [hunk 1]
```diff
@@ -4,10 +4,14 @@ import akka.actor.{Actor, ActorLogging, ActorRef, PoisonPill, Props}
...
 import io.iohk.ethereum.db.storage._
 import io.iohk.ethereum.domain.BlockHeader
+import io.iohk.ethereum.domain.Block
 import io.iohk.ethereum.network.FastSyncActor._
 import io.iohk.ethereum.network.PeerActor.MessageReceived
 import io.iohk.ethereum.network.p2p.messages.PV62._
 import io.iohk.ethereum.network.p2p.messages.PV63._
+import io.iohk.ethereum.network.p2p.validators.BlockValidator
+import io.iohk.ethereum.network.p2p.validators.BlockValidator.BlockError
+import io.iohk.ethereum.network.p2p.validators.ForkValidator
```
Comment (ntallar) [inline src/main/scala/io/iohk/ethereum/network/FastSyncActor.scala]: It seems you added some unused imports, are they needed here?

**Comment by ntallar** [comment]
Apart from that minor comment, LGTM!

File: src/main/scala/io/iohk/ethereum/network/FastSyncActor.scala (hunk 1 above)
Comment (AlanVerbner) [inline src/main/scala/io/iohk/ethereum/network/FastSyncActor.scala]: fixed

File: src/main/scala/io/iohk/ethereum/network/p2p/validators/BlockValidator.scala [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+    * @param block  Block to validate
+    * @param receipts Receipts to be in validation process
+    * @return The block if validations are ok, error otherwise
+    */
+  def validate(block: Block, receipts: Seq[Receipt]): Either[BlockError, Block] = {
+    for {
+      _ <- validateTransactionRoot(block)
+      _ <- validateOmmers(block)
+      _ <- validateReceipts(block, receipts)
+      _ <- validateLogBloom(block, receipts)
```
Comment (adamsmo) [inline src/main/scala/io/iohk/ethereum/network/p2p/validators/BlockValidator.scala]: can we also validate proof of work here?

Comment (ntallar) [inline src/main/scala/io/iohk/ethereum/network/p2p/validators/BlockValidator.scala]: I'm not sure if this answers your question but a simple proof of work validation is already implemented in `validatePoW` from `BlockHeaderValidation` (there's also a task to complete this validation https://iohk.myjetbrains.com/youtrack/issue/EC-88).

Comment (adamsmo) [inline src/main/scala/io/iohk/ethereum/network/p2p/validators/BlockValidator.scala]: yes it does, thx 👍

**Pull request merged**